### Blessy Chinthapalli

#### Individual Project 2

## Part 1: iPstack and MongoDB

In [1]:
import pymongo
from pymongo import MongoClient

# connecting to local host
conn = MongoClient("localhost", 27017)

In [2]:

# Defining database and collection name
database = "msba"
collection_1 = "ip_addresses"


db = conn[database]

# creating new collection in the db
collection = db[collection_1]

print(db.list_collection_names())

['ip_addresses']


In [3]:
# connectimg to database
database =conn['msba']
database

Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'msba')

In [4]:

# creating collection
collection = db['ip_addresses']

# Define document
document = {
    "ip": "192.168.1.1",
    "city": "Davis",
    "zip": "95616"
}

# Insert the document into collection
result = collection.insert_one(document)


print('Document ID:', result.inserted_id)

Document ID: 65f68d40f548ada5b2c03965


In [5]:
# print to check collection again after inserting document
print(db.list_collection_names())

['ip_addresses']


#### Explore IPStack

API Key: 0662f16432a967838cfb69aa1bdc69c8

In [6]:
# Specify Output Format"

# http://api.ipstack.com/134.201.250.155?access_key=0662f16432a967838cfb69aa1bdc69c8&output=json

In [7]:
# Specify Response Fields

#http://api.ipstack.com/134.201.250.155?access_key=0662f16432a967838cfb69aa1bdc69c8&fields=country_code

#### Given IP's
http://api.ipstack.com/8.8.8.8?access_key=0662f16432a967838cfb69aa1bdc69c8&output=json&fields=country_code

http://api.ipstack.com/128.120.0.25?access_key=0662f16432a967838cfb69aa1bdc69c8&output=json&fields=country_code

http://api.ipstack.com/128.32.12.14?access_key=0662f16432a967838cfb69aa1bdc69c8&output=json&fields=country_code

http://api.ipstack.com/64.165.72.144?access_key=0662f16432a967838cfb69aa1bdc69c8&output=json&fields=country_code

#### My IP address: 
73.93.241.81

In [8]:
import requests
import time

headers = {'User-Agent': 'Mozilla/5.0'}
api_key = '0662f16432a967838cfb69aa1bdc69c8'

# First IP address 
ip_address = '134.201.250.155'


url = f'http://api.ipstack.com/{ip_address}?access_key={api_key}&output=json'
#url = f'http://api.ipstack.com/134.201.250.155?access_key=0662f16432a967838cfb69aa1bdc69c8&fields=country_code'
response = requests.get(url, headers=headers)


time.sleep(10)

# Checking if req is successful
data = response.json()  # Parse to JSON
print(data)


{'ip': '134.201.250.155', 'type': 'ipv4', 'continent_code': 'NA', 'continent_name': 'North America', 'country_code': 'US', 'country_name': 'United States', 'region_code': 'CA', 'region_name': 'California', 'city': 'Los Angeles', 'zip': '90012', 'latitude': 34.0655517578125, 'longitude': -118.24053955078125, 'location': {'geoname_id': 5368361, 'capital': 'Washington D.C.', 'languages': [{'code': 'en', 'name': 'English', 'native': 'English'}], 'country_flag': 'https://assets.ipstack.com/flags/us.svg', 'country_flag_emoji': '🇺🇸', 'country_flag_emoji_unicode': 'U+1F1FA U+1F1F8', 'calling_code': '1', 'is_eu': False}}


Used the sample URL from the website and made api call, genrated the json to check results

In [9]:
response

<Response [200]>

### API CALL Program

For the Api Call, defined a function that will take the generated URL's and store the api data in a list.



In [10]:
# API CALL Program


# List of IP addresses
ip_addresses = ["8.8.8.8","128.120.0.25","128.32.12.14", "64.165.72.144", "73.93.241.81"]
api_responses = []

# function to do api calls
def API_call(ip_address):
 
 headers = {'User-Agent': 'Mozilla/5.0'}
 api_key = '0662f16432a967838cfb69aa1bdc69c8'

 url = f'http://api.ipstack.com/{ip_address}?access_key={api_key}&output=json'
 response = requests.get(url, headers=headers)
 time.sleep(5)

 data = response.json()  # Parse to JSON
 print(data)

 return response.json() 


# Make API calls 
for ip in ip_addresses:
   api_data =  API_call(ip)
   api_responses.append(api_data)

{'ip': '8.8.8.8', 'type': 'ipv4', 'continent_code': 'NA', 'continent_name': 'North America', 'country_code': 'US', 'country_name': 'United States', 'region_code': 'OH', 'region_name': 'Ohio', 'city': 'Glenmont', 'zip': '44628', 'latitude': 40.5369987487793, 'longitude': -82.12859344482422, 'location': {'geoname_id': None, 'capital': 'Washington D.C.', 'languages': [{'code': 'en', 'name': 'English', 'native': 'English'}], 'country_flag': 'https://assets.ipstack.com/flags/us.svg', 'country_flag_emoji': '🇺🇸', 'country_flag_emoji_unicode': 'U+1F1FA U+1F1F8', 'calling_code': '1', 'is_eu': False}}
{'ip': '128.120.0.25', 'type': 'ipv4', 'continent_code': 'NA', 'continent_name': 'North America', 'country_code': 'US', 'country_name': 'United States', 'region_code': 'CA', 'region_name': 'California', 'city': 'Davis', 'zip': '95616', 'latitude': 38.56296157836914, 'longitude': -121.81600952148438, 'location': {'geoname_id': 5341704, 'capital': 'Washington D.C.', 'languages': [{'code': 'en', 'name

In [11]:
api_responses

[{'ip': '8.8.8.8',
  'type': 'ipv4',
  'continent_code': 'NA',
  'continent_name': 'North America',
  'country_code': 'US',
  'country_name': 'United States',
  'region_code': 'OH',
  'region_name': 'Ohio',
  'city': 'Glenmont',
  'zip': '44628',
  'latitude': 40.5369987487793,
  'longitude': -82.12859344482422,
  'location': {'geoname_id': None,
   'capital': 'Washington D.C.',
   'languages': [{'code': 'en', 'name': 'English', 'native': 'English'}],
   'country_flag': 'https://assets.ipstack.com/flags/us.svg',
   'country_flag_emoji': '🇺🇸',
   'country_flag_emoji_unicode': 'U+1F1FA U+1F1F8',
   'calling_code': '1',
   'is_eu': False}},
 {'ip': '128.120.0.25',
  'type': 'ipv4',
  'continent_code': 'NA',
  'continent_name': 'North America',
  'country_code': 'US',
  'country_name': 'United States',
  'region_code': 'CA',
  'region_name': 'California',
  'city': 'Davis',
  'zip': '95616',
  'latitude': 38.56296157836914,
  'longitude': -121.81600952148438,
  'location': {'geoname_id': 5

In [12]:
# JSON Parsin

# print city and zip
for i in api_responses:
    city = i.get('city')
    zip_code = i.get('zip')
    print(f"City: {city}, Zip Code: {zip_code}")

City: Glenmont, Zip Code: 44628
City: Davis, Zip Code: 95616
City: Berkeley, Zip Code: 94705
City: Florin, Zip Code: 95819
City: San Francisco, Zip Code: 94103


In [13]:
# Data Storage Enhancement:

for i in api_responses:
    ip = i.get('ip')
    city = i.get('city')
    zip_code = i.get('zip')
    
    # To store results in the MongoDB ip_address collection. Defining document key values pairs
    document = {
        "ip": ip,
        "city": city,
        "zip": zip_code
    }
    
    # Insert the document into the collection
    result = collection.insert_one(document)
    
    # Printif the document is inserted
    print(f"Inserted document for IP - {ip}")

Inserted document for IP - 8.8.8.8
Inserted document for IP - 128.120.0.25
Inserted document for IP - 128.32.12.14
Inserted document for IP - 64.165.72.144
Inserted document for IP - 73.93.241.81


In [14]:
# Checking collection again after inserting document
print(db.list_collection_names())



['ip_addresses']


In [15]:
# List all documents in the ip_addresses collection
for doc in collection.find({}):
    print(doc)

{'_id': ObjectId('65f0f44c8e5af650663838aa'), 'ip': '192.168.1.1', 'city': 'Davis', 'zip': '95616'}
{'_id': ObjectId('65f145d58e5af650663838ab'), 'ip': '8.8.8.8', 'city': 'Glenmont', 'zip': '44628'}
{'_id': ObjectId('65f145d58e5af650663838ac'), 'ip': '128.120.0.25', 'city': 'Davis', 'zip': '95616'}
{'_id': ObjectId('65f145d58e5af650663838ad'), 'ip': '128.32.12.14', 'city': 'Berkeley', 'zip': '94705'}
{'_id': ObjectId('65f145d58e5af650663838ae'), 'ip': '64.165.72.144', 'city': 'Florin', 'zip': '95819'}
{'_id': ObjectId('65f145d58e5af650663838af'), 'ip': '73.93.241.81', 'city': 'San Francisco', 'zip': '94103'}
{'_id': ObjectId('65f68d40f548ada5b2c03965'), 'ip': '192.168.1.1', 'city': 'Davis', 'zip': '95616'}
{'_id': ObjectId('65f68dbcf548ada5b2c03966'), 'ip': '8.8.8.8', 'city': 'Glenmont', 'zip': '44628'}
{'_id': ObjectId('65f68dbcf548ada5b2c03967'), 'ip': '128.120.0.25', 'city': 'Davis', 'zip': '95616'}
{'_id': ObjectId('65f68dbcf548ada5b2c03968'), 'ip': '128.32.12.14', 'city': 'Berkele

## Part 2: eBay + Selenium

In [47]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains
import time

driver = webdriver.Chrome()

driver.implicitly_wait(10)
driver.set_script_timeout(120)
driver.set_page_load_timeout(10)

#driver.get('https://google.com')
driver.get('https://www.ebay.com/')


time.sleep(3)


input = driver.find_element(By.XPATH, "//input[@placeholder='Search for anything']")
input.send_keys('Cell Phones\n')
#time.sleep(10)
time.sleep(5)


time.sleep(5)

# filter network
filters = driver.find_element(By.XPATH, f"//span[contains(text(), 'Unlocked')]")
filters.click()
time.sleep(3)

# scroll after first filter
filters = driver.find_elements(By.XPATH, f"//span[contains(text(), 'Network')]/following::span[contains(text(), 'Unlocked')]")
if filters:
        # Scroll to the filter checkbox
        driver.execute_script("arguments[0].scrollIntoView(true);", filters[0])
        ActionChains(driver).move_to_element(filters[0]).perform()
time.sleep(3)


# Filter Brand
filters = driver.find_element(By.XPATH, f"//span[contains(text(), 'LG')]")
filters.click()
time.sleep(3)

# scroll after LG filter
filters = driver.find_elements(By.XPATH, f"//span[contains(text(), 'Brand')]/following::span[contains(text(), 'LG')]")
if filters:
        # Scroll to the filter checkbox
        driver.execute_script("arguments[0].scrollIntoView(true);", filters[0])
        ActionChains(driver).move_to_element(filters[0]).perform()
time.sleep(3)



# Filter Screen Size
filters = driver.find_element(By.XPATH, f"//span[contains(text(), '6 in or More')]")
filters.click()
time.sleep(3)


# scroll after scrren size filter
filters = driver.find_elements(By.XPATH, f"//span[contains(text(), 'Screen Size')]/following::span[contains(text(), '6 in or More')]")
if filters:
        # Scroll to the filter checkbox
        driver.execute_script("arguments[0].scrollIntoView(true);", filters[0])
        ActionChains(driver).move_to_element(filters[0]).perform()
time.sleep(3)

# Filter Storage Capacity
filters = driver.find_element(By.XPATH, f"//span[contains(text(), '128 GB')]")
filters.click()
time.sleep(3)

# Filter nStorage Capacity
filters = driver.find_element(By.XPATH, f"//span[contains(text(), 'Factory Unlocked')]")
filters.click()
time.sleep(3)

# filter Condition
filters = driver.find_element(By.XPATH, f"//span[contains(text(), 'Used')]")
filters.click()
time.sleep(3)

# save to html file
with open("unlocked-lg-128gb-used.html", "w", encoding="utf-8") as file:
    file.write(driver.page_source)


driver.quit()

NOTE: The filter selection works even without scrolling also. So stopped scrolling after 3rd selection.

In [46]:
from bs4 import BeautifulSoup
import re

# open html file
with open("unlocked-lg-128gb-used.html", "r", encoding="utf-8") as file:
    soup = BeautifulSoup(file, "html.parser")
    #soup.prettify



# iterate through each item listing
for i in soup.select('.s-item'):
    # To extract title
    title = i.select_one('.s-item__title').text 
    
    # To extract seller info and percentage (rating)
    seller = i.select_one('.s-item__seller-info-text').text if i.select_one('.s-item__seller-info-text') else "N/A"
    
    # Extract seller rating - using regex
    rating = re.search(r'(\d+(\.\d+)?)%', seller)

    # If rating is there, extract the percentage
    if rating: seller_rating = rating.group(1)
    else: seller_rating = "N/A"

    # To extract the price
    price = i.select_one('.s-item__price').text 
    
    # Shipping information  
    shipping = i.find("span", class_="s-item__shipping s-item__logisticsCost").text if i.find("span", class_="s-item__shipping s-item__logisticsCost") else "Shipping information - Not available"
   
    # Returns information
    returns = i.find("span", class_="s-item__shipping s-item__logisticsCost").text if i.find("span", class_="s-item__shipping s-item__logisticsCost") else "Shipping information - Not available "
   

    # Number of bids 
    bids = i.find("span", class_="s-item__dynamic s-item__quantitySold").text if i.find("span", class_="s-item__dynamic s-item__quantitySold") else "Bidding Info - Not Available"
   
    

    print(f"Title: {title}")
    print(f"Seller: {seller}")
    print(f"Seller Rating (in %): {seller_rating}")
    print(f"Price: {price}")

    print(f"Shipping Information: {shipping}")
    print(f"Return Information: {returns}")

    print(f"Number of bids: {bids}")
    print("-------\n")  


Title: Shop on eBay
Seller: N/A
Seller Rating (in %): N/A
Price: $20.00
Shipping Information: Shipping information - Not available
Return Information: Shipping information - Not available 
Number of bids: Bidding Info - Not Available
-------

Title: LG Velvet 5G 128GB Grey LM-G900TM (T-Mobile Unlocked) Reduced Price zW7656
Seller: soonersoft (159,723) 98.8%
Seller Rating (in %): 98.8
Price: $104.19
Shipping Information: Shipping information - Not available
Return Information: Shipping information - Not available 
Number of bids: Bidding Info - Not Available
-------

Title: LG Velvet 5G - LM-G900VMP 128GB Gray Verizon + GSM Unlocked LTE Smartphone GREAT
Seller: discountphonedepot (4,272) 97.6%
Seller Rating (in %): 97.6
Price: $116.73
Shipping Information: Free shipping
Return Information: Free shipping
Number of bids: Bidding Info - Not Available
-------

Title: LG K61 LG-Q630UM 128GB Unlocked White Android Smartphone
Seller: monkeytote (17,816) 98.5%
Seller Rating (in %): 98.5
Price: 

With only one filter - Network Unlocked.

To print all the listing information

In [50]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains
import time

driver = webdriver.Chrome()

driver.implicitly_wait(10)
driver.set_script_timeout(120)
driver.set_page_load_timeout(10)

#driver.get('https://google.com')
driver.get('https://www.ebay.com/')


time.sleep(3)


input = driver.find_element(By.XPATH, "//input[@placeholder='Search for anything']")
input.send_keys('Cell Phones\n')
#time.sleep(10)
time.sleep(5)


filters = driver.find_element(By.XPATH, f"//span[contains(text(), 'Unlocked')]")
filters.click()

time.sleep(5)

# save to html file
with open("unlocked.html", "w", encoding="utf-8") as file:
    file.write(driver.page_source)


driver.quit()

In [49]:
from bs4 import BeautifulSoup
import re

# open html file
with open("unlocked.html", "r", encoding="utf-8") as file:
    soup = BeautifulSoup(file, "html.parser")
    #soup.prettify



# iterate through each item listing
for i in soup.select('.s-item'):
    # To extract title
    title = i.select_one('.s-item__title').text 
    
    # To extract seller info and percentage (rating)
    seller = i.select_one('.s-item__seller-info-text').text if i.select_one('.s-item__seller-info-text') else "N/A"
    
    # Extract seller rating - using regex
    rating = re.search(r'(\d+(\.\d+)?)%', seller)

    # If rating is there, extract the percentage
    if rating: seller_rating = rating.group(1)
    else: seller_rating = "N/A"

    # To extract the price
    price = i.select_one('.s-item__price').text 
    
    # Shipping information  
    shipping = i.find("span", class_="s-item__shipping s-item__logisticsCost").text if i.find("span", class_="s-item__shipping s-item__logisticsCost") else "Shipping information - Not available"
   
    # Returns information
    returns = i.find("span", class_="s-item__shipping s-item__logisticsCost").text if i.find("span", class_="s-item__shipping s-item__logisticsCost") else "Shipping information - Not available "
   

    # Number of bids 
    bids = i.find("span", class_="s-item__dynamic s-item__quantitySold").text if i.find("span", class_="s-item__dynamic s-item__quantitySold") else "Bidding Info - Not Available"
   
    

    print(f"Title: {title}")
    print(f"Seller: {seller}")
    print(f"Seller Rating (in %): {seller_rating}")
    print(f"Price: {price}")

    print(f"Shipping Information: {shipping}")
    print(f"Return Information: {returns}")

    print(f"Number of bids: {bids}")
    print("-------\n")  


Title: Shop on eBay
Seller: N/A
Seller Rating (in %): N/A
Price: $20.00
Shipping Information: Shipping information - Not available
Return Information: Shipping information - Not available 
Number of bids: Bidding Info - Not Available
-------

Title: Samsung Galaxy S21 5G Unlocked G991U 128GB Android Smartphone Excellent
Seller: bigmoosewireless (13,895) 99.2%
Seller Rating (in %): 99.2
Price: $228.99
Shipping Information: Free shipping
Return Information: Free shipping
Number of bids: 348+ sold
-------

Title: 😍New SEALED Samsung Galaxy S20 5G SM-G981U1 128GB FACTORY UNLOCKED Cell Phones
Seller: elec-ml (1,828) 99.1%
Seller Rating (in %): 99.1
Price: $306.99
Shipping Information: Free shipping
Return Information: Free shipping
Number of bids: Bidding Info - Not Available
-------

Title: OnePlus 10T 5G 128GB Moonstone Black  T-Mobile Unlocked Smartphone - Brand New.
Seller: amazing-wireless (135,973) 99.4%
Seller Rating (in %): 99.4
Price: $284.99
Shipping Information: Shipping informat

Note: 
The first listing is generated out of no where. 
However, from second listing its exact as the website.

Title: Samsung Galaxy S21 5G Unlocked G991U 128GB Android Smartphone Excellent

Seller: bigmoosewireless (13,895) 99.2%

Seller Rating (in %): 99.2

Price: $228.99

Shipping Information: Free shipping

Return Information: Free shipping

Number of bids: 348+ sold


![alt text](image-1.png)